synapse/storage/databases/main/event_federation.py

```diff
@@ -1130,6 +1197,21 @@ def _delete_old_forward_extrem_cache_txn(txn):
             _delete_old_forward_extrem_cache_txn,
         )
 
+    async def insert_insertion_extremity(self, event_id: str, room_id: str) -> None:
+        def _insert_insertion_extremity_txn(txn):
+            self.db_pool.simple_insert_txn(
```

Comment:
You can use `await self.db_pool.simple_insert(...)` helper, to avoid this boilerplate

Reply:
Also, we probably want to use a `simple_upsert` here, to handle the case where we've already processed the event somehow, I think?

Reply:
Should I be preferring the `async` helpers over the `txn` stuff? Any context behind why both?

It seems some of the code isn't `async` yet and kinda forces into a `txn` corner.

Reply:
> Should I be preferring the `async` helpers over the `txn` stuff? Any context behind why both?

Yes, as it reduces the boiler plate. We have `_txn` versions for when we want to run multiple queries in a single transaction (for atomicity).

> It seems some of the code isn't `async` yet and kinda forces into a `txn` corner.

The async helpers should  always be usable wherever we're doing this currently, as `runInteraction` is also `async`. What you may be seeing is that if you have a function that calls a single async function then technically it can just be a normal function that returns the result of the call to the async function (i.e. just return the coroutine, rather than awaiting it and returning the result). That's not a great thing to do though, as it messes up stack traces, but they crept in as we converted stuff from `@defer.inlineCallbacks` to async/await.

Reply:
Thanks for the info 😀

I'm thinking specifically of [`synapse/storage/databases/main/events.py`](https://github.com/matrix-org/synapse/blob/72935b7c5053af122c8cb5767a3e85a3a0f3a20c/synapse/storage/databases/main/events.py#L1826) where I had to refactor `get_room_version` to be available as `get_room_version_txn`. But it seems like "We have _txn versions for when we want to run multiple queries in a single transaction (for atomicity)." applies in this case so we can persist multiple events at once (`_persist_events_txn`).

Reply:
> But it seems like "We have _txn versions for when we want to run multiple queries in a single transaction (for atomicity)." applies in this case so we can persist multiple events at once (`_persist_events_txn`).

fwiw, it also applies when persisting only one event, as we want to either add the event to all the appropriate tables or none of them.